src/libraries/System.Linq/src/System/Linq/Cast.cs

```diff
@@ -55,5 +61,90 @@ private static IEnumerable<TResult> CastIterator<TResult>(IEnumerable source)
                 yield return (TResult)obj;
             }
         }
+
+        [DebuggerDisplay("Count = {Count}")]
+        private sealed partial class CastIListIterator<TResult>(IList source) : Iterator<TResult>, IList<TResult>, IReadOnlyList<TResult>
+        {
+            private readonly IList _source = source;
+            private IEnumerator? _enumerator;
+
+            public override Iterator<TResult> Clone() => new CastIListIterator<TResult>(_source);
+
+            public override bool MoveNext()
+            {
+                switch (_state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        _state = 2;
+                        goto case 2;
+
+                    case 2:
+                        Debug.Assert(_enumerator != null);
+                        if (_enumerator.MoveNext())
+                        {
+                            _current = (TResult)_enumerator.Current;
+                            return true;
+                        }
+
+                        Dispose();
+                        break;
+                }
+
+                return false;
+            }
+
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    (_enumerator as IDisposable)?.Dispose();
+                    _enumerator = null;
+                }
+
+                base.Dispose();
+            }
+
+            public int Count => _source.Count;
+
+            public bool IsReadOnly => true;
+
+            public int IndexOf(TResult item) => _source.IndexOf(item);
+
+            public bool Contains(TResult item) => _source.Contains(item);
+
+            public void CopyTo(TResult[] array, int arrayIndex)
+            {
+                ArgumentNullException.ThrowIfNull(array);
+
+                IList source = _source;
+                int count = source.Count;
+
+                if (arrayIndex < 0 || arrayIndex > array.Length - count)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+                }
+
+                for (int i = 0; i < count; i++)
+                {
+                    array[arrayIndex + i] = (TResult)source[i]!;
+                }
+            }
+
+            public TResult this[int index]
```

Comment:
If you exclude the indexer, this looks like an optimization that could be extended to `ICollection` sources in general. Wouldn't it make sense to target that broader type instead? (And perhaps keep a derived IList iterator type for lists)

Reply:
Does that buy us enough? All arrays and `List<T>` will implement `IList<T>`, which represents a large swath of what we'd care about here. We can't implement `ICollection<T>` on the resulting type, because with just `ICollection` we wouldn't have an efficient way to implement `ICollection<T>.Contains` (we'd need to enumerate the `ICollection` via the non-generic enumerator). Without an indexer, we couldn't efficiently implement First/Last. Basically the only thing we'd be able to do more efficiently would be implement `IPartition` to expose a count and pre-size ToArray/List.

Reply:
That said, it looks like this whole IList approach has issues with multidimensional arrays, which implement IList but which don't like being indexed into via the indexer.  So, I may just drop down to ICollection and we can live with the lesser benefits from implementing IPartition instead.